File: pubsubsystem/publisher.go
package pubsubsystem

import "fmt"

type Publisher struct {
	Topics map[*Topic]struct{}
}

func NewPublisher() *Publisher {
	return &Publisher{Topics: make(map[*Topic]struct{})}
}

func (p *Publisher) RegisterTopic(topic *Topic) {
	p.Topics[topic] = struct{}{}
}

func (p *Publisher) Publish(topic *Topic, message *Message) {
	if _, exists := p.Topics[topic]; !exists {
		fmt.Printf("No usch topic exists in the publisher:: %s\n", topic.Name)
	}

	topic.PublishMsg(message)
}

File: pubsubsystem/typescript/publisher.ts
import { Topic } from '../topic';
import { Message } from './message';

export class Publisher {
    topics: Map<Topic, boolean>;

    constructor() {
        this.topics = new Map<Topic, boolean>();
    }

    static newPublisher(): Publisher {
        return new Publisher();
    }

    registerTopic(topic: Topic): void {
        this.topics.set(topic, true);
    }

    publish(topic: Topic, message: Message): void {
        if (!this.topics.has(topic)) {
            console.log(`No such topic exists in the publisher: ${topic.name}`);
            return;
        }

        topic.publishMsg(message);
    }
}